api/caseshare/real-api.ts

```diff
@@ -46,65 +48,111 @@ export async function getCases(req: EnhancedRequest, res: Response, next: NextFu
   }
 }
 
-export async function assignCases(req: EnhancedRequest, res: Response, next: NextFunction) {
+export async function assignCases(req: EnhancedRequest, res: Response, next: NextFunction): Promise<Response> {
   const shareCases: SharedCase[] = req.body.sharedCases.slice()
 
-  // call share case api
-  const {updatedSharedCases, updatedErrorMessages}:
-    { updatedSharedCases: SharedCase[], updatedErrorMessages: string[] } = await doShareCase(req, shareCases)
+  let allPromises
 
-  // TODO: call unshare case api
-  // await doUnshareCase(req, shareCases, updatedSharedCases)
+  // call share case api(the n call)
+  const shareCasePromises = doShareCase(req, shareCases)
+  allPromises = [...shareCasePromises]
+  // call unShare case api(the 1 batch call)
+  const unShareCasePromise = doUnShareCase(req, shareCases)
+  if (unShareCasePromise !== null) {
+    allPromises = [...shareCasePromises, unShareCasePromise]
+  }
+
+  // @ts-ignore
+  const allResults = await Promise.allSettled(allPromises)
+
+  const { updatedErrorMessages,
+          rejectedShareCasePayloads,
+          rejectedUnShareCaseReason,
+          rejectedCount } = handleRejectedResponse(allResults)
 
-  const originalSharedNumber = shareCases.reduce((acc, aCase) => acc
-    + (aCase.pendingShares ? aCase.pendingShares.length : 0), 0)
-  const afterSharedNumber = updatedSharedCases.reduce((acc, aCase) => acc
-    + (aCase.pendingShares ? aCase.pendingShares.length : 0), 0)
+  const updatedSharedCases: SharedCase[] = handleSharedCase(shareCases, rejectedShareCasePayloads)
 
-  // when none of the users are assigned successfully
-  if (originalSharedNumber > 0 && originalSharedNumber === afterSharedNumber) {
+  const finalSharedCases = handleUnSharedCase(updatedSharedCases, rejectedUnShareCaseReason)
+  // when none of the apis successfully
+  if (allPromises.length === rejectedCount) {
     return res.status(500).send(updatedErrorMessages)
+  } else {
+    return res.status(201).send(finalSharedCases)
   }
-
-  // when all/partial are assigned successfully
-  return res.status(201).send(updatedSharedCases)
 }
 
-async function doShareCase(req: EnhancedRequest, shareCases: SharedCase[]):
-  Promise<{updatedSharedCases: SharedCase[], updatedErrorMessages: string[]}> {
+function doShareCase(req: EnhancedRequest, shareCases: SharedCase[]): any[] {
   const path = `${ccdUrl}/case-assignments`
-  const promises = promiseCaseShareBatchCall(shareCases, path, req)
+  const promises = []
   // @ts-ignore
```

Comment:
_Question:_
Given this related to the `await Promise.allSettled(promises)` call, is this still needed? Right now, it'll ignore the `shareCases.flatMap(...)` call.

Reply:
yes it's needed, same as await Promise.allSettled(promises), coz the typescript version in node layer is low, it can't recognize flatMap, compliation will be failed without @ts-ignore.  If bump up the node version it needs full regression test on node.